### PR TITLE
Implement `Vector.build`: implementation and some tests

### PR DESCRIPTION
**Pull request by GregoryTravis**
This PR implements `Vector.build` and converts uses of `.new_builder` to `.build` within Vector and Vector_Spec.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso [hunk 1]
```diff
@@ -171,6 +174,50 @@ type Vector a
     new_builder : Integer -> Builder
     new_builder (capacity=10) = Builder.new capacity
 
+    ## PRIVATE
+       ADVANCED
+       Creates a new `Vector` by providing a `Builder` to the provided function.
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso]: I'd rephrase this sentence as _providing X to the provided Y_ sounds weird.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso [hunk 2]
```diff
@@ -171,6 +174,50 @@ type Vector a
...
+       > Example
+         Construct a vector using a builder that contains the items 1 to 5.
+
+             Vector.build builder->
+                 builder.append 1
+                 builder.append 2
+                 builder.append 3
+             # => [1, 2, 3]
+    build : (Builder -> Any) -> Integer -> Vector
+    build (function : Builder -> Any) (capacity : Integer = 10) -> Vector =
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso]:
```suggestion
    build (function : Builder -> Any) (initial_capacity : Integer = 10) -> Vector =
```

I think maybe adding `initial_` would clarify the meaning of this variable; without _initial_ it could sound like it's the _max_ capacity.

File: test/Base_Tests/src/Data/Vector_Spec.enso [hunk 3]
```diff
@@ -1034,6 +1030,26 @@ add_specs suite_builder =
...
+            v2 = Vector.build builder->
+                _ = builder  # Do nothing
+            v2 . should_equal []
+
+        group_builder.specify "Should propagate dataflow errors" <|
+            v = Vector.build builder->
+                builder.append 1
+                builder.append 2
+                Error.throw (Illegal_Argument.Error "asdf")
+            v . should_fail_with (Illegal_Argument.Error "asdf")
```
Comment (radeusgd) [inline test/Base_Tests/src/Data/Vector_Spec.enso:1055]: This is a great test.

Not directly related, but while we are at it we could consider solving a long outstanding problem we used to have:
```
        group_builder.specify "Should propagate the first dataflow error encountered in `append`" <|
            v = Vector.build builder->
                builder.append 1
                builder.append (Error.throw (Illegal_Argument.Error "ERROR2")
                builder.append 3
                builder.append (Error.throw (Illegal_Argument.Error "ERROR4")
                builder.append 5
            v . should_fail_with (Illegal_Argument.Error "ERROR2")
```

Unfortunately, as far as I know currently this will actually not fail but produce a vector `[1, 3, 5]`, essentially swallowing the dataflow errors.

We used to have a pattern of 'folding' the builder i.e. `append` returns a builder and this would have to be rewritten as:
```
                builder . append 1
                        . append (Error.throw (Illegal_Argument.Error "ERROR2")
                        . append 3
                        . append (Error.throw (Illegal_Argument.Error "ERROR4")
                        . append 5
```
in which case the error would be correctly propagated.

Unfortunately, the 'folding the builder' pattern did not get much traction. In many places it was just not too practical - the builder was inherently mutable object so this more 'functional' folding was often not fitting its usages.

Do you think we could amend the builder to detect that `append` got a dataflow error and avoid swallowing it?

This could be implemented by holding a `encountered_error : Ref (Any | Nothing)` inside of `Builder`, then `append` can check `item.is_error` and set this `Ref` the first time an error is encountered; then `Builder.to_vector` would first check if `encountered_error` is set and if so, propagate that error instead of building the vector.

Comment (radeusgd) [reply test/Base_Tests/src/Data/Vector_Spec.enso:1055]: But this is a bit out of scope of this PR, so I guess we can also do it as a separate ticket/PR instead.

**Review by radeusgd — APPROVED** [review]
Looks great.

Your test inspired me to try improving a dataflow-error swallowing issue with `Builder.append` - but I guess that may be a task for a separate PR.

File: test/Base_Tests/src/Data/Vector_Spec.enso (hunk 3 above)
Comment (GregoryTravis) [reply test/Base_Tests/src/Data/Vector_Spec.enso:1055]: Very good idea. #9733

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso (hunk 1 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso]: Changed to "passing a `Builder` to the provided function"

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso (hunk 2 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso]: Done

**Review by jdunkerley — APPROVED** [review]

File: test/Base_Tests/src/Data/Vector_Spec.enso [hunk 4]
```diff
@@ -778,21 +776,19 @@ type_spec suite_builder name alter = suite_builder.group name group_builder->
 
     group_builder.specify "should correctly propagate state through each" <|
         v = State.run Number 77 <|
-            b = Vector.new_builder
-            (alter ["A", "B"]).each x->
-                b.append x
-                b.append (State.get Number).to_text
-            b.to_vector
+            Vector.build builder->
```
Comment (JaroslavTulach) [inline test/Base_Tests/src/Data/Vector_Spec.enso:779]: I'd assume this is a general even for complicated computations. Just wrap your computation in `Vector.build builder->` block and that's it.
